controllers/nutanixcluster_controller.go

```diff
@@ -289,20 +321,94 @@ func (r *NutanixClusterReconciler) reconcileNormal(rctx *nctx.ClusterContext) (r
 
 func (r *NutanixClusterReconciler) reconcileFailureDomains(rctx *nctx.ClusterContext) error {
 	log := ctrl.LoggerFrom(rctx.Context)
-	if len(rctx.NutanixCluster.Spec.FailureDomains) == 0 {
-		log.V(1).Info("no failure domains defined on cluster")
-		conditions.MarkTrue(rctx.NutanixCluster, infrav1.NoFailureDomainsReconciled)
+	log.Info("Reconciling failure domains for cluster")
+
+	// First clear the failure domain related conditions
+	conditions.Delete(rctx.NutanixCluster, infrav1.FailureDomainsValidatedCondition)
+	conditions.Delete(rctx.NutanixCluster, infrav1.NoFailureDomainsConfiguredCondition)
+
+	failureDomains := capiv1.FailureDomains{}
+
+	if len(rctx.NutanixCluster.Spec.ControlPlaneFailureDomains) == 0 && len(rctx.NutanixCluster.Spec.FailureDomains) == 0 {
+		log.Info("No failure domains configured for cluster.")
+		conditions.MarkTrue(rctx.NutanixCluster, infrav1.NoFailureDomainsConfiguredCondition)
+		// Reset the failure domains for nutanixcluster status
+		rctx.NutanixCluster.Status.FailureDomains = failureDomains
 		return nil
 	}
-	log.V(1).Info("Reconciling failure domains for cluster")
-	// If failure domains is nil on status object, first create empty slice
-	if rctx.NutanixCluster.Status.FailureDomains == nil {
-		rctx.NutanixCluster.Status.FailureDomains = make(capiv1.FailureDomains, 0)
+
+	// Not found failure domain names
+	notFoundFDs := []string{}
+	// Map of mis-configured failure domain name and validation error message
+	misConfigFDs := map[string]string{}
+
+	for _, fdRef := range rctx.NutanixCluster.Spec.ControlPlaneFailureDomains {
+		// Fetch the referent failure domain object
+		fdObj := &infrav1.NutanixFailureDomainV2{}
+		fdKey := client.ObjectKey{Name: fdRef.Name, Namespace: rctx.NutanixCluster.Namespace}
+		err := r.Client.Get(rctx.Context, fdKey, fdObj)
+		if err != nil {
+			log.Error(err, fmt.Sprintf("Failed to fetch the referent failure domain object %q.", fdRef.Name))
+			notFoundFDs = append(notFoundFDs, fdRef.Name)
```

Comment:
this doesn't necessarily mean that the failure domain was not found. 

please check to see if the error is specifically is an `IsNotFound` error.

Reply:
The notFoundFDs include those the fetching call of the FD failed, no matter what the specific error is. Because to move to the next step of validating the FD configuration, the FD object is required to fetch without any error.

Reply:
Reword the error message.